### PR TITLE
Fixes matplotlib/matplotlib#1235

### LINKED ISSUE #1235: Legend placement bug
This script produces a legend that covers the red point:

```python
import matplotlib.pyplot as plt

colors = ['b','g','r']
for n in range(3):
    plt.scatter([n,],[n,],color=colors[n])
plt.legend(['foo','foo','foo'],loc='best')
plt.gca().set_xlim(-0.5, 2.2)
plt.gca().set_ylim(-0.5, 2.2)
plt.show()
```

### PR DESCRIPTION
**Pull request by nansonzheng**
Credit to @dashed, fixes #1235, supercedes #2857 

The problem is that _find_best_position does not receive positions of (transformed) points from _auto_legend_data. We attempted to fix it by adding the offsets of points, after preparing the points and transforming them, and calling legendBox.count_contains on offsets.

### CODE REVIEW

**Comment by dashed** [comment]
AppVeyor builds are failing; but I cannot make sense of the logs. Can someone enlighten me?

**Comment by tacaswell** [comment]
See #6115 That is probably unrelated to your work

File: lib/matplotlib/legend.py [hunk 1]
```diff
@@ -737,6 +737,8 @@ def _auto_legend_data(self):
         ax = self.parent
         bboxes = []
         lines = []
+        vertices = []
```
Comment (tacaswell) [inline lib/matplotlib/legend.py]: Do not need this, it looks like vertices just gets re-assigned below.

**Comment by tacaswell** [comment]
Other that 1 very minor style point :+1:

File: lib/matplotlib/tests/test_legend.py [hunk 2]
```diff
@@ -261,6 +262,31 @@ def test_nanscatter():
...
+
+    for n in range(3):
+        plt.scatter([n,], [n,], color=colors[n])
+
+    plt.legend(['foo', 'foo', 'foo'], loc='best')
+    plt.gca().set_xlim(-0.5, 2.2)
+    plt.gca().set_ylim(-0.5, 2.2)
+
+
+@image_comparison(baseline_images=['not_covering_scatter_transform'], extensions=['png'])
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_legend.py]: I think this line is over 80 characters.

Comment (dashed) [inline lib/matplotlib/tests/test_legend.py]: Done.

I noticed the rest of the test file violates this :wink: Can be amended in a separat PR.

File: lib/matplotlib/legend.py (hunk 1 above)
Comment (dashed) [inline lib/matplotlib/legend.py]: Fixed.

File: lib/matplotlib/tests/test_legend.py (hunk 2 above)
Comment (tacaswell) [inline lib/matplotlib/tests/test_legend.py]: I am fine putting off fixing pep8 in the tests.

Comment (QuLogic) [inline lib/matplotlib/tests/test_legend.py]: It's actually fixed in another PR that I might just give a little nudge to some time...

**Comment by dashed** [comment]
@tacaswell I rebased against master that has https://github.com/matplotlib/matplotlib/pull/6116, and seems to work. Unsure what happened here: https://ci.appveyor.com/project/mdboom/matplotlib/build/1.0.985/job/4u3na3cignvxhg2n

**Comment by tacaswell** [comment]
That is a known transient failure.

**Comment by tacaswell** [comment]
backported to v2.x as 869cc9d

**Comment by dashed** [comment]
Thanks @tacaswell  :+1:

**Pull request merged**